Readme.md

```diff
@@ -0,0 +1,1432 @@
+
+```scala
+case object EmptyList
+```
+
+Second, an ADT may combine several mutually exclusive variants of products:
+
+```scala
+sealed trait OptionalInt
+case class SomeInt(value: Int) extends OptionalInt
+case object NoneInt extends OptionalInt
+```
+
+Here `OptionalInt` is the ADT, and `SomeInt` and `NoneInt` are its variants. Each variant is a product ADT in itself. Such combinations of product types are called sum types, or sometimes disjoint unions. Another common example of such an ADT is a tree, which also demonstrates how an ADT can be recursive:
+
+```scala
+sealed trait Tree
+object Tree {
+  case class Node(left: Tree, right: Tree) extends Tree
+  case class Leaf(value: Int) extends Tree
+}
+```
+
+As can be seen above, in Scala sealed traits and case classes are used to create algebraic data types.
+
+ADTs are useful to model a wide range data structures, simple data type and state machines being the most prominent examples. Pattern matching together with the exhaustiveness analysis allow writing more correct code. Many algorithms on ADTs are naturally recursive, and pattern matching leads to very clear and correct code:
+
+```scala
+def findMax(tree: Tree): Int = tree match {
+  case Tree.Node(left, right) => Vector(findMax(left), findMax(right)).max
+  case Tree.Leaf(value) => value
+}
+```
+
+Because of the exhaustiveness analysis, such pattern matches will make sure that you didn't forget to check any variants of an ADT.
+
+#### Sum types declarations
+
+If you need to write a sum data type, declare all its variants inside the companion object of the root sealed trait:
+
+```scala
+// write this
+sealed trait IntList
+object IntList {
+  case class Cons(value: Int, rest: IntList) extends IntList
+  case object Empty extends IntList
+}
+
+// not this
+sealed trait IntList
+case class IntListCons(value: Int, rest: IntList) extends IntList
+case object IntListEmpty extends IntList
+```
+
+Putting variants of a sum type into the companion object reduces the namespace pollution, allows to use nicer names for variants and makes the code which uses these names clearer because it becomes visible to which type these variants belong.
+
+#### Enumerations
+
+One important case of sum ADTs are enumerations, i.e. a fixed small set of values (objects). Lots of things are modeled with enumerations, for example, state of some system:
+
+```scala
+sealed trait State
+object State {
+  case object Stopped extends State
+  case object Started extends State
+  case object Failed extends State
+}
+```
+
+So, when you need to declare something which would be an `enum` in Java, use the above pattern.
+
+Just as with any other ADT, declare all enumeration constants inside the companion object of the respective sealed trait. Do not use `scala.Enumeration` because it has very unnatural API.
+
+### Options and `null`
```

Comment:
> marking optional fields in serialized objects in some way is not so bad

Well, these fields are likely to be used by clients in languages which don't have an explicit notion of an optional value (e.g. Javascript), as well as in languages which do not support fancy syntax like `_?`. And both in these languages and in the data format, as opposed to static languages like Scala, absence of a field is not modeled as a special value, it is modeled as, well, absence of the field. So there is no need to mark these fields in any special way there because it is not really useful.

And since we don't do it in DTOs, then for uniformity reasons I would be very much against adopting a different practice in the other parts of the system.